In [3]:
import numpy as np

What is a Neural Network?

The funtion neural_network is a representation of a neural network in python. It takes in inputs and weights and multiplies each data point to predict a particular value. 

What is Input data?
It is a number that is recorded in the real world and it usually somethign that is knowable, it can be tooday's temperature. or yesterday's stock price or the image of a cat. 

What is a Prediction?
A prediction is what the neural informs is based on the provided input such as "given yesterday's Microsoft stock price, the stock price will be 150.52".

Predictions are not always right, sometime the neural networks makes mistakes and it can also learn from these mistakes, if it sees the prediction was weither too high or too low, it makes changes the weigh either up or down based on the predictions results so as to predict more accurately next time. 


How does the Neural Network work?
Basically, it mulplies the input by a weight, that is it scales the input by a certail anount. The weight is a number that is learned by the neural network. The neural network is trained on a dataset, and it learns the weights that will give it the best results. The weights are adjusted based on the difference between the predicted value and the actual value. 

In [4]:
# simple neural network making a prediction on if a team would win a game
# based on the number of toes of the players in the team

# function contains an empty network as no input is provided

weight = 0.1
def neural_network(input, weight):
    prediction = input * weight 
    return prediction 

In [5]:
# inserting one input datapoint to the function 
num_of_toes = [8.5, 9.5, 10, 9]
input = num_of_toes[0]
pred = neural_network(input, weight)
print(pred)

0.8500000000000001


# Making a prediction with multiple inputs

now we want to provide muliple inputs to the function tomake a prediction
since toes are not the only factors that can detremine if players would play well and 
win a game we would add toes, wlrec, nfans
toes = courrent average number of toes per player
wlrec = current games won (%)
nfans = fan count in million
we would then multiply each of these inputs by a weight to get a prediction

The prediction is 0.98% which shows that adding more data or context to the neural network
hekps improve the prediction results compared to when we had only toes as input which gave us a 
prediction of 0.85%


The new property introced here is that we have multiple inputs and we have to sum their 
respective predictions. Therefore, we multiply each input byt its respective weight and then sum all
the precitions tigether and this called *Weighted Sum of the Input or Weighted Sum*. We can also refer to it as 
*Dot Product*. 

Example: 
inputs are [toes, wlrec, nfans]
Here we would multiply the inputs values in index 1 their corresponding weight value 
toes[1] = 9.5 * 0.1 = 0.95
wlrec[1] = 0.8 * 0.2 = 0.16
nfans[1] = 1.3 * 0.0 = 0.00

Then we sum all to get index 1 prediction
0.95 + 0.16 + 0.00 = 1.11

In [6]:
# now we want to provide muliple inputs to the function tomake a prediction
# since toes are not the only factors that can detremine if players would play well and
weights = [0.1, 0.2, 0]
def neural_network(input, weight):
    prediction = w_sum(input * weights) 
    return prediction 

In [7]:
print(type(input))
print(type(weights))

<class 'float'>
<class 'list'>


In [8]:


toes = [8.5, 9.5, 9.9, 9.0]
wlrec = [0.65, 0.8, 0.8, 0.9]
nfans = [1.2, 1.3, 0.5, 1.0]

input = [toes[0], wlrec[0], nfans[0]]
weights = [0.1, 0.2, 0]


def w_sum(a, b):
    assert(len(a) == len(b))
    output = 0.0
    for i in range(len(a)):
        output += (a[i] * b[i])
    return output

def neural_network(input, weight):
    prediction = w_sum(input, weights) 
    return prediction 

pred = neural_network(input, weights)
print(pred)



0.9800000000000001


In [9]:

print(pred)

0.9800000000000001


This new need to process multiple inputs at a time justifies the use of a new tool.is called a *Vector*. it. A vector is nothing other than a list of numbers. In the example, input is a vector and weights is a vector. 

Anytime you perform a mathematical operation between two vectors of equal length where you pair up values according to their position in the vector (again: position 0 with 0, 1 with 1, and so on), it’s called an elementwise operation. Thus elementwise addition sums two vectors, and elementwise multiplication multiplies two vectors.


## Challenge: Vector math

Being able to manipulate vectors is a cornerstone technique for deep learning. See if you can
write functions that perform the following operations:
• def elementwise_multiplication(vec_a, vec_b)
• def elementwise_addition(vec_a, vec_b)
• def vector_sum(vec_a)
• def vector_average(vec_a)
Then, see if you can use two of these methods to perform a dot product!

In [10]:
# weights = [0.1, 0.2, 0.3]
toes = [8.5, 9.5, 9.9]
def elementwise_multiplication(a, b):
    assert(len(a) == len(b))
    output = []
    for i in range(len(a)):
        output.append(a[i] * b[i])
    return output


def elementwise_addition(a, b):
    assert(len(a) == len(b))
    output = []
    for i in range(len(a)):
        output.append(a[i] + b[i])
    return output

def vector_sum(a):
    output = 0.0
    for i in range(len(a)):
        output += a[i]
    return output

def vector_mean(a):
    output = vector_sum(a)
    return round(output / len(a), 2)


toes_mult = elementwise_multiplication(toes, weights)
toes_add = elementwise_addition(toes, weights)
toes_sum = vector_sum(toes)
toes_mean = vector_mean(toes)

print("Elementwise Multiplication:", toes_mult)
print("Elementwise Addition:", toes_add)
print("Vector Sum:", toes_sum)
print("Vector Mean:", toes_mean)

Elementwise Multiplication: [0.8500000000000001, 1.9000000000000001, 0.0]
Elementwise Addition: [8.6, 9.7, 9.9]
Vector Sum: 27.9
Vector Mean: 9.3


In [11]:
inputsweights = elementwise_multiplication(input, weights)
vectorsums = vector_sum(inputsweights)
print(inputsweights)
print(vectorsums)

[0.8500000000000001, 0.13, 0.0]
0.9800000000000001


In [12]:
# numpy implementation of multiple inputs to neural network 

weights = np.array([0.1, 0.2, 0])
def neural_network(input, weights):
    # dot function in numpy represents the dot product which earlier we used w_sum function
    pred = input.dot(weights)
    return pred


toes = np.array([8.5, 9.5, 9.9, 9.0])
wlrec = np.array([0.65, 0.8, 0.8, 0.9])
nfans = np.array([1.2, 1.3, 0.5, 1.0])

input = np.array([toes[0], wlrec[0], nfans[0]])
pred = neural_network(input, weights)
print(pred)

0.9800000000000001


### Note: 
Weights cant be shuffled, the weights simply means this is how much a particular feature contributes to the prediction of something. For our example the most sensitive weight is the wlrec with 0.2 and number of fans has the least with 0 therefore ist results are basically negligble in how the team performs. Even as wlrec has the most sensitive weight, the most dominant force in the high score is the number of toes not because the weight is the highest but because the input combined with the weight provides the highest results so far. 


# Making predictions with Multiple Outputs
Neural networks can make multiple predictions using only one single input value. Predictions occurs the same as if there were three disconnected single weight neural networks. Using our example but this time around we would tweak it to see players emotions based on if they won or lost. So we want to predict three values hurt, happy or sad, we want to see the percentage of the teams who feel these emotions 

weights = [0.3, 0.2, 0.9]
the first weight is for hurt, second for happy and the last for sad

we would use the same values of our wlrec variable 



In [13]:
weights = [0.3, 0.2, 0.9]

def neural_network(input, weights):
    pred = ele_mul(input, weights)
    return pred

wlrec = [0.65, 0.8, 0.8, 0.9]
input = wlrec[0]


def ele_mul(number, vectors):
    output = [0, 0 ,0]
    assert(len(output) == len(vectors))
    for i in range(len(output)):
        output[i] = number * vectors[i]
    
    return output

def neural_network(input, weights):
    pred = ele_mul(input, weights)
    return pred

pred = neural_network(input, weights)
print(pred)

[0.195, 0.13, 0.5850000000000001]


### Translating Results

- 0.195 = hurt prediction 
- 0.13 = happy prediction
- 0.585 = sad prediction

## Multiple Inputs for Multiple Outputs

Neural networks can predict multiple outputs givcen multiple inputs. A weight connect each input node to each output node and the prediction is going to occur the usual manner. 

### How does this work?
Previously we were working with one weights for each feature but here, instead of having one set of weights (for one output), you have multiple sets of weights — one for each output. This means you're doing multiple dot products, one for each output neuron.

Example 

input = [9.5, 0.8, 1.3]

weights = [
    [0.1, 0.1, -0.3],  # weights for output neuron 1
    [0.1, 0.2, 0.0],   # weights for output neuron 2
    [0.0, 1.3, 0.1]    # weights for output neuron 3
]


### Math Breakdown
We want to calculate:

output= output=input * weight1, input * weight2, input * weight3
​

​
 
Output 1= (9.5×0.1)+(0.8×0.1)+(1.3×−0.3)
=(9.5×0.1)+(0.8×0.1)+(1.3×−0.3)
=0.95+0.08−0.39= 0.64
​
Output 2 = (9.5×0.1)+(0.8×0.2)+(1.3×0.0)
=(9.5×0.1)+(0.8×0.2)+(1.3×0.0)
=0.95+0.16+0.0= 1.11
​
 
Output 3 = (9.5×0.0)+(0.8×1.3)+(1.3 + 0.1)
=(9.5×0.0)+(0.8×1.3)+(1.3×0.1)
=0.0+1.04+0.13= 
1.17
​

Here: Inputs are represented as vectors and weights as matrix, so we are having a vector_matrix multiplication, where each input is multiplied by a row in the matrix. The result is a vector of outputs.


In [14]:

weights =[[ 0.1, 0.1, -0.3],
          [0.1, 0.2, 0.0],
          [0.0, 1.3, 0.1]]

def neural_network(input, weights):
    pred = vect_mat_mul(input, weights)
    return pred


toes = [8.5, 9.5, 9.9, 9.0]
wlrec = [0.65, 0.8, 0.8, 0.9]
nfans = [1.2, 1.3, 0.5, 1.0]
input = [toes[0], wlrec[0], nfans[0]]
print(len(input))
print(len(weights[0]))

def w_sum(a, b):
    assert(len(a) == len(b))
    output = 0.0
    for i in range(len(a)):
        output += (a[i] * b[i])
    return output

def vect_mat_mul(vect, matrix):
    assert(len(vect) == len(matrix))
    output = [0, 0, 0]
    for i in range(len(vect)):
        output[i] = w_sum(vect, matrix[i])
    return output
    
def neural_network(input, weights):
    pred = vect_mat_mul(input, weights)
    return pred

pred = neural_network(input, weights)
print(pred)

3
3
[0.555, 0.9800000000000001, 0.9650000000000001]


# Neural Networks can be stacked
What does this mean? 
An output from one layer can be used as an input to another layer. This is called a stacked neural network. 

In [ ]:
# Lets implement a stacked neural network with numpy 
ih_wgt = np.array([
                #toes, wlrec, nfans
              [0.1,0.2, -0.1], #hid[0]
              [-0.1, 0.1, 0.9], #hid[1]
                [0.1, 0.4, 0.1]]) #hid[
 
hp_wgt = np.array([
    #hid[0], hid[1], hid[2]
    [0.3, 1.1, -0.3], # hurt?
    [0.1, 0.2, 0.0], # win?
    [0.0, 1.3, 0.1]  # sad?
])

weights = [ih_wgt, hp_wgt]
def neural_network(input, weights):
    hid = input.dot(weights[0]) 
    pred = hid.dot(weights[1]) 
    return pred
toes = np.array([8.5, 9.5, 9.9, 9.0]) 
wlrec = np.array([0.65,0.8, 0.8, 0.9]) 
nfans = np.array([1.2, 1.3, 0.5, 1.0])
input = np.array([toes[0],wlrec[0],nfans[0]])
pred = neural_network(input,weights) 
print(pred)

[ 0.496  1.256 -0.286]


Here all we have discussed is a forl of Forward Propagation, where a neural network takes input data and makes a prediction. It is called Forward Propagation because we are propagating activations forward through the network. In our examples, activations are all the numbers that are not weights and are unique for every prediction. 